chain/client/src/sync.rs

```diff
@@ -400,6 +400,49 @@ impl BlockSync {
             _ => chain.head()?.last_block_hash,
         };
 
+        let reference_hash = {
+            // Find the most recent block we known on the canonical chain.
+            // In practice the forks from the last final block are very short, so it is
+            // acceptable to perform this on each request
+            let header = chain.get_block_header(&reference_hash)?;
+            let mut candidate = (header.height(), *header.hash(), *header.prev_hash());
+
+            // First go back until we find the common block
+            while match chain.get_header_by_height(candidate.0) {
+                Ok(header) => header.hash() != &candidate.1,
+                Err(e) => match e.kind() {
+                    near_chain::ErrorKind::DBNotFoundErr(_) => true,
+                    _ => return Err(e),
+                },
+            } {
+                let prev_header = chain.get_block_header(&candidate.2)?;
+                candidate = (prev_header.height(), *prev_header.hash(), *prev_header.prev_hash());
+            }
+
+            // Then go forward for as long as we known the next block
+            let mut ret_hash = candidate.1;
+            loop {
+                match chain.mut_store().get_next_block_hash(&ret_hash) {
+                    Ok(hash) => {
+                        let hash = hash.clone();
+                        if chain.block_exists(&hash)? {
+                            ret_hash = hash;
```

Comment:
```suggestion
                        if chain.block_exists(hash)? {
                            ret_hash = hash.clone();
```

Reply:
I applied the change, but will undo it, a reference to hash keeps a reference on `chain`, and cannot be used in `chain.block_exists`.